File: app/SourcesShared/IPTVConverterClient.swift
import Foundation

/// Client for the hosted IPTV converter at `iptv.vortx.tv`. It registers a user's M3U playlist or Xtream Codes
/// login with the worker, which stores the (encrypted) credentials server-side and returns an opaque slug; the
/// app then installs `https://iptv.vortx.tv/c/<slug>/manifest.json` as a normal Stremio add-on so the channels
/// flow through the existing Live tab engine-side. Removing a playlist calls /revoke so the slug is destroyed.
///
/// GATING: `iptv.vortx.tv` is in `VortXEdgeAuth.gatedHosts`, so every request here is HMAC-signed. Signing is a
/// safe no-op without a provisioned secret (the worker runs the gate in OBSERVE mode), so this works today and
/// tightens later without a client change.
enum IPTVConverterClient {

    /// A user-facing error from a register / revoke attempt.
    enum ClientError: LocalizedError {
        case badResponse
        case xtreamAuthFailed
        case server(String)
        case network(String)

        var errorDescription: String? {
            switch self {
            case .badResponse: return String(localized: "The IPTV service returned an unexpected response.")
            case .xtreamAuthFailed: return String(localized: "Those Xtream details did not sign in. Check the server, username, and password.")
            case .server(let code): return String(localized: "The IPTV service could not add this playlist (\(code)).")
            case .network(let msg): return msg
            }
        }
    }

    /// A successful registration: the worker slug and the manifest URL to install.
    struct Registration {
        let slug: String
        let manifestURL: String
    }

    /// The base URL of the converter worker.
    private static var baseURL: URL { URL(string: "https://iptv.vortx.tv")! }

    // MARK: - Register

    /// Register an M3U playlist. `xmltvURL` is an optional separate EPG source.
    static func registerM3U(url: String, xmltvURL: String?, name: String?) async -> Result<Registration, ClientError> {
        var body: [String: Any] = ["m3u_url": url]
        if let xmltvURL, !xmltvURL.isEmpty { body["xmltv_url"] = xmltvURL }
        if let name, !name.isEmpty { body["name"] = name }
        return await register(body: body)
    }

    /// Register an Xtream Codes login. The worker validates the credentials before returning a slug.
    static func registerXtream(host: String, user: String, pass: String, xmltvURL: String?, name: String?) async -> Result<Registration, ClientError> {
        var body: [String: Any] = ["xtream": ["host": host, "user": user, "pass": pass]]
        if let xmltvURL, !xmltvURL.isEmpty { body["xmltv_url"] = xmltvURL }
        if let name, !name.isEmpty { body["name"] = name }
        return await register(body: body)
    }

    private static func register(body: [String: Any]) async -> Result<Registration, ClientError> {
        guard let data = try? JSONSerialization.data(withJSONObject: body) else {
            return .failure(.badResponse)
        }
        var req = URLRequest(url: baseURL.appendingPathComponent("register"), timeoutInterval: 20)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "content-type")
        req.httpBody = data
        VortXEdgeAuth.sign(&req)

        do {
            let (respData, resp) = try await URLSession.shared.data(for: req)
            guard let http = resp as? HTTPURLResponse else { return .failure(.badResponse) }
            let obj = (try? JSONSerialization.jsonObject(with: respData)) as? [String: Any]
            if (200..<300).contains(http.statusCode) {
                guard let obj, let slug = obj["slug"] as? String, let manifestURL = obj["manifest_url"] as? String else {
                    return .failure(.badResponse)
                }
                return .success(Registration(slug: slug, manifestURL: manifestURL))
            }
            // Non-2xx: surface the worker's error code, mapping the known auth failure to a friendly message.
            let code = (obj?["error"] as? String) ?? String(http.statusCode)
            if code == "xtream_auth_failed" { return .failure(.xtreamAuthFailed) }
            return .failure(.server(code))
        } catch {
            return .failure(.network(error.localizedDescription))
        }
    }

    // MARK: - Revoke

    /// Revoke a slug server-side so its playlist can no longer be served. Fail-soft: a failure here is logged
    /// but does not block local removal (the add-on is uninstalled either way). Returns true on a clean revoke.
    @discardableResult
    static func revoke(slug: String) async -> Bool {
        guard !slug.isEmpty else { return false }
        var req = URLRequest(url: baseURL.appendingPathComponent("c").appendingPathComponent(slug).appendingPathComponent("revoke"),
                             timeoutInterval: 12)
        req.httpMethod = "POST"
        VortXEdgeAuth.sign(&req)
        do {
            let (_, resp) = try await URLSession.shared.data(for: req)
            return (resp as? HTTPURLResponse).map { (200..<300).contains($0.statusCode) } ?? false
        } catch {
            return false
        }
    }
}

File: android/app/src/main/kotlin/com/vortx/android/iptv/IPTVConverterClient.kt
package com.vortx.android.iptv

import org.json.JSONObject

/// Client for the hosted IPTV converter at `iptv.vortx.tv`. Kotlin port of the Apple
/// `app/SourcesShared/IPTVConverterClient.swift`, with the SAME worker endpoints and the SAME request /
/// response shapes (derived from the Swift client, not invented).
///
/// It registers a user's M3U playlist or Xtream Codes login with the worker, which stores the (encrypted)
/// credentials server-side and returns an opaque slug; the app then installs
/// `https://iptv.vortx.tv/c/<slug>/manifest.json` as a normal add-on so the channels flow through the
/// existing catalog / Live pipeline engine-side. Removing a playlist calls `/c/<slug>/revoke` so the slug is
/// destroyed server-side, not just forgotten locally.
///
/// GATING: `iptv.vortx.tv` is in [com.vortx.android.net.VortXEdgeAuth]'s gated-host set, so every request
/// here is HMAC-signed by [IPTVHttp] (a safe no-op without a provisioned secret, so this works today and
/// tightens later without a client change), mirroring the Apple `VortXEdgeAuth.sign(&req)`.
///
/// The request-shape builders ([registerBodyM3U] / [registerBodyXtream]), the path builders, and the result
/// mapper ([mapRegisterResult]) are pure and `internal` so they can be unit-tested without the network.
object IPTVConverterClient {

    /// A user-facing error from a register attempt. Extends [Exception] so it rides Kotlin's [Result] failure
    /// channel (the repository idiom), while [message] carries copy matching the Apple `ClientError`.
    sealed class ClientError(message: String) : Exception(message) {
        /// The service returned a non-JSON / malformed body or a 2xx without a slug + manifest URL.
        object BadResponse : ClientError("The IPTV service returned an unexpected response.")

        /// The worker's `xtream_auth_failed` code: the Xtream credentials did not sign in.
        object XtreamAuthFailed :
            ClientError("Those Xtream details did not sign in. Check the server, username, and password.")

        /// Any other non-2xx: the worker's error code (or the raw status when it sent none).
        class Server(val code: String) : ClientError("The IPTV service could not add this playlist ($code).")

        /// A transport failure (offline / DNS / TLS / timeout): no HTTP response reached the client.
        object Network : ClientError("Could not reach the IPTV service. Check your connection and try again.")
    }

    /// The base URL of the converter worker (Apple `IPTVConverterClient.baseURL`).
    const val BASE_URL = "https://iptv.vortx.tv"

    // MARK: - Register

    /// Register an M3U playlist. [xmltvUrl] is an optional separate EPG source; [name] an optional label.
    /// Mirrors Apple `registerM3U(url:xmltvURL:name:)`.
    suspend fun registerM3U(url: String, xmltvUrl: String?, name: String?): Result<IPTVRegistration> =
        register(registerBodyM3U(url, xmltvUrl, name))

    /// Register an Xtream Codes login (the worker validates the credentials before returning a slug).
    /// Mirrors Apple `registerXtream(host:user:pass:xmltvURL:name:)`.
    suspend fun registerXtream(
        host: String,
        user: String,
        pass: String,
        xmltvUrl: String?,
        name: String?,
    ): Result<IPTVRegistration> = register(registerBodyXtream(host, user, pass, xmltvUrl, name))

    private suspend fun register(body: Map<String, Any?>): Result<IPTVRegistration> {
        val response = IPTVHttp.request(
            method = "POST",
            urlString = "$BASE_URL/$REGISTER_PATH",
            body = bodyToJson(body),
            timeoutMs = IPTVHttp.REGISTER_TIMEOUT_MS,
        )
        val obj = parseJsonObject(response.body)
        return mapRegisterResult(
            status = response.status,
            slug = obj?.optString("slug")?.ifEmpty { null },
            manifestUrl = obj?.optString("manifest_url")?.ifEmpty { null },
            errorCode = obj?.optString("error")?.ifEmpty { null },
        )
    }

    // MARK: - Revoke

    /// Revoke a slug server-side so its playlist can no longer be served. Fail-soft: a failure here is not
    /// surfaced (the add-on is uninstalled either way). Returns true on a clean revoke. Mirrors Apple
    /// `revoke(slug:)`.
    suspend fun revoke(slug: String): Boolean {
        if (slug.isEmpty()) return false
        val response = IPTVHttp.request(
            method = "POST",
            urlString = "$BASE_URL/${revokePath(slug)}",
            timeoutMs = IPTVHttp.REVOKE_TIMEOUT_MS,
        )
        return response.isSuccess
    }

    // MARK: - Pure request-shape + result seams (unit-tested)

    /// The worker's register path (Apple `baseURL.appendingPathComponent("register")`).
    internal const val REGISTER_PATH = "register"

    /// The worker's revoke path for a slug (Apple `baseURL/c/<slug>/revoke`). The slug is a worker-generated
    /// opaque capability token, so it is interpolated as a single already-safe path segment.
    internal fun revokePath(slug: String): String = "c/$slug/revoke"

    /// The M3U register body: `{"m3u_url": url}`, with `xmltv_url` and `name` added ONLY when non-empty
    /// (matches the Apple `registerM3U` body construction exactly). A [LinkedHashMap] so key order is stable
    /// for tests; the wire order is irrelevant to the worker.
    internal fun registerBodyM3U(url: String, xmltvUrl: String?, name: String?): LinkedHashMap<String, Any?> {
        val body = LinkedHashMap<String, Any?>()
        body["m3u_url"] = url
        if (!xmltvUrl.isNullOrEmpty()) body["xmltv_url"] = xmltvUrl
        if (!name.isNullOrEmpty()) body["name"] = name
        return body
    }

    /// The Xtream register body: `{"xtream": {"host": host, "user": user, "pass": pass}}`, with `xmltv_url`
    /// and `name` added ONLY when non-empty (matches the Apple `registerXtream` body construction exactly).
    internal fun registerBodyXtream(
        host: String,
        user: String,
        pass: String,
        xmltvUrl: String?,
        name: String?,
    ): LinkedHashMap<String, Any?> {
        val xtream = LinkedHashMap<String, Any?>()
        xtream["host"] = host
        xtream["user"] = user
        xtream["pass"] = pass
        val body = LinkedHashMap<String, Any?>()
        body["xtream"] = xtream
        if (!xmltvUrl.isNullOrEmpty()) body["xmltv_url"] = xmltvUrl
        if (!name.isNullOrEmpty()) body["name"] = name
        return body
    }

    /// Map a worker response (status + the fields the network path extracted) to the typed result. Pure, so
    /// the whole decision table is unit-tested without a socket. Mirrors the Apple `register` result branch:
    ///   - status 0 (transport failure) -> [ClientError.Network]
    ///   - 2xx with slug + manifest URL  -> success
    ///   - 2xx missing either field      -> [ClientError.BadResponse]
    ///   - non-2xx `xtream_auth_failed`  -> [ClientError.XtreamAuthFailed]
    ///   - any other non-2xx            -> [ClientError.Server] (the worker's code, or the raw status)
    internal fun mapRegisterResult(
        status: Int,
        slug: String?,
        manifestUrl: String?,
        errorCode: String?,
    ): Result<IPTVRegistration> {
        if (status == 0) return Result.failure(ClientError.Network)
        if (status in 200..299) {
            if (slug.isNullOrEmpty() || manifestUrl.isNullOrEmpty()) return Result.failure(ClientError.BadResponse)
            return Result.success(IPTVRegistration(slug = slug, manifestUrl = manifestUrl))
        }
        val code = errorCode?.takeIf { it.isNotEmpty() } ?: status.toString()
        if (code == "xtream_auth_failed") return Result.failure(ClientError.XtreamAuthFailed)
        return Result.failure(ClientError.Server(code))
    }

    // MARK: - JSON (network path only)

    /// Serialize an ordered body map to a JSON string, converting one level of nested map (the `xtream`
    /// object) into a nested `JSONObject`. Kept off the pure builders above so the request-shape tests never
    /// need `org.json` on the classpath.
    private fun bodyToJson(body: Map<String, Any?>): String {
        val root = JSONObject()
        for ((key, value) in body) {
            when (value) {
                null -> {}
                is Map<*, *> -> {
                    val nested = JSONObject()
                    for ((k, v) in value) if (v != null) nested.put(k.toString(), v)
                    root.put(key, nested)
                }
                else -> root.put(key, value)
            }
        }
        return root.toString()
    }

    private fun parseJsonObject(text: String): JSONObject? =
        if (text.isEmpty()) null else runCatching { JSONObject(text) }.getOrNull()
}
